llvm/lib/Transforms/Scalar/ConstraintElimination.cpp

```diff
@@ -1736,6 +1765,11 @@ static bool eliminateConstraints(Function &F, DominatorTree &DT, LoopInfo &LI,
         Changed |= Simplified;
       } else if (auto *MinMax = dyn_cast<MinMaxIntrinsic>(Inst)) {
         Changed |= checkAndReplaceMinMax(MinMax, Info, ToRemove);
+      } else if (auto *CmpIntrinsic = dyn_cast<IntrinsicInst>(Inst);
+                 CmpIntrinsic &&
+                 (CmpIntrinsic->getIntrinsicID() == Intrinsic::scmp ||
+                  CmpIntrinsic->getIntrinsicID() == Intrinsic::ucmp)) {
```

Comment:
Do we need a helper class like `CmpIntrinsic`?
```
class CmpIntrinsic : public IntrinsicInst {
public:
  static bool classof(const IntrinsicInst *I) {...}
  Value *getLHS() const {...}
  Value *getRHS() const {...}
  bool isSigned() {...}
  static bool isSigned(Intrinsic::ID ID) {...}
  ICmpInst::Predicate getEqPredicate() {...}
  static ICmpInst::Predicate getEqPredicate(Intrinsic::ID ID) {...}
  ICmpInst::Predicate getLtPredicate() {...}
  static ICmpInst::Predicate getLtPredicate(Intrinsic::ID ID) {...}
  ICmpInst::Predicate getGtPredicate() {...}
  static ICmpInst::Predicate getGtPredicate(Intrinsic::ID ID) {...}
};
```

Reply:
Would be good to have a common helper to handle both ICMp and those intrinsics together, at least in ConstraintElimination to reduce duplication if possible

Reply:
Added the suggested class. I can't think of any good way to remove code duplication between `icmp` and `ucmp/scmp`, but if there are any suggestions I'm happy to implement them.

Reply:
> Added the suggested class.

Please file a separate NFC PR for this change.

Reply:
Done in #98177